draft-ietf-tls-esni.md

```diff
@@ -482,6 +469,39 @@ would divulge the true server name. The remaining
 contents of the ClientHelloOuter MAY be identical to those in
 ClientHelloInner but MAY also differ.
 
+## Recommended Padding Scheme {#padding}
+
+This section describes a deterministic padding mechanism based on the following
+observation: individual extensions can reveal sensitive information through their
+length. Thus, each extension in the inner ClientHello may require different amounts
+of padding. This padding may be fully determined by the client's configuration or
+may require server input.
+
+By way of example, a client's set of ALPN values is typically fixed. Clients SHOULD
+pad this extension by rounding up to the longest of those values. The target padding
+length of most ClientHello extensions can be computed in this way.
+
+In contrast, clients do not know the longest SNI value in the client-facing server's
+anonymity set without server input. For the "server_name" extension with length D,
+clients SHOULD use the server's length hint L (ECHOCOnfig.maximum_name_length) when
+computing the padding as follows:
+
+1. If L > D, add L - D bytes of padding. This rounds to the server's advertised
+hint, i.e., ECHOCOnfig.maximum_name_length.
+2. Otherwise, add 32 - (D % 32) bytes of padding. This rounds D up to the nearest
+multiple of 32 bytes.
+
+The amount of padding applied to the ClientHello is computed as the sum of
+all per-extension padding values, rounded up to the nearest multiple of 32.
+(This additional rounding step aims to hide variance across different client
+implementations.)
```

Comment:
Say we just padded SNI and L = 0 and D = 10. We then compute the per-extension padding value as 22, but because of this paragraph we pad by 32? If so, I don't think this works, because if D was 11, we would also pad by 32, so now we are leaking the length of D.

Reply:
Discussed offline -- will remove and consider later on, as this exists to mask fixed-length values across different stacks, and fingerprinting stacks is already pretty trivial.